### PR TITLE
feat(settings): wire extension env vars to pi-config-settings.json

### PR DESCRIPTION
**Pull request by myakove**
## Summary
- Adds five extension settings to pi-config-settings.json: pidash_enable, pidiff_enable, pidash_port, image_model, acpx_agents
- Resolution order: project file > global ~/.pi/pi-config-settings.json > OS env var > default (backward compatible)
- Wires settings into pidash, pidiff, async-runner, image-gen, acpx-provider
- Renames review-state module and JSON data file to a unique name to block CLI bypass of the review enforcement system
- Hardens rules/20-code-review-loop.md with commit enforcement documentation and warnings against review state manipulation
- Makes extension-settings tests deterministic via setGlobalSettingsPath (isolates from real ~/.pi)
- Adds hasGitAddForce enforcement helper to block git add with force flag (respects -- and shell separators)
- Fixes 0ms reviewer results leaving review state stuck in pending (bounded retry after 3 failures)

## Test plan
- [x] npx tsx --test tests/node/**/*.test.ts (347 tests)
- [x] uv run --group tests pytest (172 passed)
- [x] prek run --all-files
- [ ] Reload pi and verify /pidash status respects pidash_port from settings
- [ ] Verify acpx_agents in settings registers acpx models without ACPX_AGENTS env

### CODE REVIEW

File: extensions/orchestrator/project-settings.ts [hunk 1]
```diff
@@ -45,6 +51,19 @@ function parseSettingsFile(filePath: string): ProjectSettings {
     if (typeof raw.dco === "boolean") result.dco = raw.dco;
     if (typeof raw.comment_signature === "boolean") result.comment_signature = raw.comment_signature;
     if (typeof raw.review_loop_enforcement === "boolean") result.review_loop_enforcement = raw.review_loop_enforcement;
+    if (typeof raw.pidash_enable === "boolean") result.pidash_enable = raw.pidash_enable;
+    if (typeof raw.pidiff_enable === "boolean") result.pidiff_enable = raw.pidiff_enable;
+    if (typeof raw.pidash_port === "number" && Number.isFinite(raw.pidash_port)) {
+      result.pidash_port = raw.pidash_port;
+    }
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/project-settings.ts]: Fixed: parseSettingsFile now validates pidash_port with Number.isInteger() and range 1-65535. getSetting also has defensive validation. Added tests for invalid ports (0, -1, 70000, 19190.5).

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/project-settings.ts [hunk 2]
```diff
@@ -180,6 +237,43 @@ export function getSetting(cwd: string, key: string): boolean | string | number
       if (env !== undefined) return env;
       return false; // default: disabled (opt-in)
     }
+    case "acpx_agents": {
+      if (projectSettingsFileHasKey(cwd, "acpx_agents")) {
+        return parseAcpxAgentList(loadProjectSettings(cwd).acpx_agents);
+      }
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/project-settings.ts:260]: Fixed: acpx_agents getSetting now checks projectValue !== undefined before returning, so malformed types fall through to global/env. parseAcpxAgentList trims array entries. Added tests for malformed values and whitespace.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/project-settings.ts:56` (qodo bug) — Invalid pidash_port accepted
> Addressed: Fixed: parseSettingsFile now validates pidash_port with Number.isInteger() and range 1-65535. getSetting has defensive double-validation. Tests added for edge cases.
### `extensions/orchestrator/project-settings.ts:240` (qodo bug) — acpx_agents override misparses
> Addressed: Fixed: getSetting acpx_agents checks projectValue !== undefined after parseSettingsFile. Invalid types skip to global/env fallback. parseAcpxAgentList trims entries. Tests cover malformed project values.
### `tests/node/orchestrator/extension-settings.test.ts:110` (qodo finding) — Test names include and
> Addressed: Fixed: test names renamed to remove and - split into single-behavior tests.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/node/orchestrator/extension-settings.test.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+	it("pidash_port from env when project omits it", () => {
+		process.env.PI_PIDASH_PORT = "8888";
+		clearSettingsCache();
+		if (globalSettingsHasKey("pidash_port")) {
+			assert.equal(typeof getSetting(tmp, "pidash_port"), "number");
+			return;
+		}
+		assert.equal(getSetting(tmp, "pidash_port"), 8888);
+	});
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/node/orchestrator/extension-settings.test.ts:108]: Fixed: added setGlobalSettingsPath() to isolate tests from real ~/.pi/. All env-fallback tests are now deterministic with no conditional branching on global config.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `tests/node/orchestrator/extension-settings.test.ts:110` (qodo finding) — Env-fallback tests are nondeterministic
> Addressed: Fixed: added setGlobalSettingsPath() to isolate tests from real ~/.pi/. All env-fallback tests are now deterministic with no conditional branching on global config.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement.ts [hunk 4]
```diff
@@ -189,6 +189,14 @@ async function checkGitProtection(command: string, event: any, ctx: any, gitCwd:
     };
   }
 
+  // Block git add --force / -f — bypasses .gitignore and allows staging ignored files
+  if (hasGitSub(command, "add") && (/\bgit\b.*\badd\b.*(?:\s-f\b|\s--force\b)/.test(command))) {
+    return {
+      block: true,
+      reason: "⛔ 'git add -f/--force' forbidden. It bypasses .gitignore and stages ignored files.",
+    };
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement.ts]: Fixed: regex broadened to catch combined short options like -fn via -\S*f pattern.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/project-settings.ts [hunk 5]
```diff
@@ -56,8 +75,35 @@ function loadProjectSettings(cwd: string): ProjectSettings {
   return parseSettingsFile(getSettingsPath(cwd));
 }
 
+/** Override global settings path for testing. When set, loadGlobalSettings reads from this instead of ~/.pi/. */
+let globalSettingsPathOverride: string | null = null;
+
+/** Set a custom global settings path (for tests). Pass null to restore default. */
+export function setGlobalSettingsPath(path: string | null): void {
+  globalSettingsPathOverride = path;
+}
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/project-settings.ts:85]: Fixed: setGlobalSettingsPath now calls clearSettingsCache() immediately.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/enforcement.ts:192` (qodo bug) — git add -f bypass
> Addressed: Fixed: regex broadened to catch combined short options like -fn via -\S*f pattern.
### `extensions/orchestrator/project-settings.ts:78` (qodo bug) — Global override cache stale
> Addressed: Fixed: setGlobalSettingsPath now calls clearSettingsCache() immediately.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement.ts [hunk 6]
```diff
@@ -189,6 +189,14 @@ async function checkGitProtection(command: string, event: any, ctx: any, gitCwd:
     };
   }
 
+  // Block git add --force / -f (including combined short options like -fn) — bypasses .gitignore
+  if (hasGitSub(command, "add") && (/\bgit\b.*\badd\b.*(?:\s-\S*f|\s--force\b)/.test(command))) {
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement.ts]: Fixed: replaced regex with hasGitAddForce() helper that parses tokens and respects -- end-of-options marker. Added 8 tests covering -f, --force, -fn, -vf, -- -f pathspec, --pathspec-from-file.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/enforcement.ts:192` (qodo bug) — Force-flag regex false positives
> Addressed: Fixed: replaced regex with hasGitAddForce() helper that parses tokens and respects -- end-of-options marker. Added 8 tests covering -f, --force, -fn, -vf, -- -f pathspec, --pathspec-from-file.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement-helpers.ts [hunk 7]
```diff
@@ -332,6 +332,22 @@ export function isRmInProjectTmp(stmt: string, cwd: string): boolean {
...
+  if (!hasGitSub(command, "add")) return false;
+  const addMatch = command.match(/\bgit\b.*\badd\b\s+(.*)/);
+  if (!addMatch) return false;
+  const tokens = addMatch[1].split(/\s+/);
+  for (const token of tokens) {
+    if (token === "--") break; // Everything after -- is a pathspec
+    if (token === "--force") return true;
+    // Short option: -f or combined like -fn, -vf (starts with - but not --)
+    if (token.startsWith("-") && !token.startsWith("--") && token.includes("f")) return true;
+  }
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts:350]: Fixed: hasGitAddForce now splits on shell separators before tokenizing. Added tests for compound commands.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/async-agents.ts [hunk 8]
```diff
@@ -371,8 +375,13 @@ export function registerAsyncAgents(
...
           const findings = countFindings(output);
-          // -1 means invalid JSON output — treat conservatively as having findings
-          recordReviewerResult(jobCwd(job), job.agent, findings < 0 ? 1 : findings);
+          // Skip recording if the reviewer never actually ran (0ms + no valid output).
+          // This leaves the reviewer in pending state so it gets re-spawned.
+          if (findings < 0 && (!data.durationMs || data.durationMs < 100)) {
+            asyncLog(`Reviewer ${job.agent} returned no valid output (${data.durationMs}ms) — skipping recordReviewerResult to keep it pending`);
+          } else {
+            recordReviewerResult(jobCwd(job), job.agent, findings < 0 ? 1 : findings);
+          }
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/async-agents.ts]: Fixed: bounded retry (3 attempts). After 3 consecutive fast failures, records as 1 finding to unblock review state.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/async-agents.ts:378` (qodo bug) — Reviewer pending can stick
> Addressed: Fixed: bounded retry (3 attempts). After 3 consecutive fast failures, records as 1 finding to unblock review state.
### `extensions/orchestrator/enforcement-helpers.ts:337` (qodo bug) — Force check misparses compounds
> Addressed: Fixed: hasGitAddForce now splits on shell separators before tokenizing. Added tests for compound commands.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/async-agents.ts [hunk 9]
```diff
@@ -262,11 +262,16 @@ export function registerAsyncAgents(
...
+    // before processResultFile() has read all group members' outputs.
+    // Wait up to 2s for missing result files before delivering with empty data.
     const lateIngestedIds = new Set<string>();
     for (const j of groupJobs) {
       if (j.output !== undefined) continue; // Already ingested
       const rp = path.join(ASYNC_RESULTS_DIR, `${j.id}.json`);
+      // Wait briefly for the result file if it doesn't exist yet
+      for (let wait = 0; wait < 4 && !fs.existsSync(rp); wait++) {
+        await new Promise(r => setTimeout(r, 500));
+      }
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/async-agents.ts]: Fixed: replaced per-job serial wait with single group-level deadline (2s total). All missing files polled concurrently.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/async-agents.ts:266` (qodo bug) — Serial wait inflates latency
> Addressed: Fixed: replaced per-job serial wait with single group-level deadline (2s total). All missing files polled concurrently.
### `extensions/orchestrator/async-agents.ts:271` (qodo finding) — deliverGroupResults wait lacks test
> Addressed: Fixed: added tests/node/orchestrator/async-agents-group-wait.test.ts with 4 tests covering immediate find, delayed file, deadline expiry, and single-deadline verification.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/node/orchestrator/async-agents-group-wait.test.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+				found.add(id);
+			} else {
+				allFound = false;
+			}
+		}
+		if (allFound) break;
+		await new Promise((r) => setTimeout(r, 100));
+	}
+	return found;
+}
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/node/orchestrator/async-agents-group-wait.test.ts]: Fixed: extracted waitForResultFiles into async-wait.ts. Test imports and exercises the real production helper.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/async-agents.ts [hunk 11]
```diff
@@ -262,10 +262,20 @@ export function registerAsyncAgents(
...
     const lateIngestedIds = new Set<string>();
-    for (const j of groupJobs) {
-      if (j.output !== undefined) continue; // Already ingested
+    const missingJobs = groupJobs.filter(j => j.output === undefined);
+    if (missingJobs.length > 0) {
+      const deadline = Date.now() + 2000;
+      while (Date.now() < deadline) {
+        const stillMissing = missingJobs.filter(j => j.output === undefined && !fs.existsSync(path.join(ASYNC_RESULTS_DIR, `${j.id}.json`)));
+        if (stillMissing.length === 0) break;
+        await new Promise(r => setTimeout(r, 250));
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/async-agents.ts]: Fixed: missingJobs now filters out fireAndForget jobs.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/node/orchestrator/async-agents-group-wait.test.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+	it("uses single deadline for all files, not per-file", async () => {
+		tmp = mkdtempSync(join(tmpdir(), "async-group-wait-"));
+		const start = Date.now();
+		// 3 missing files, 500ms deadline — should take ~500ms total, not 1500ms
+		const found = await waitForMissingFiles(tmp, ["a", "b", "c"], 500);
+		const elapsed = Date.now() - start;
+		assert.equal(found.size, 0);
+		assert.ok(elapsed < 1000, `Expected <1000ms but took ${elapsed}ms (serial wait bug)`);
+		rmSync(tmp, { recursive: true, force: true });
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/node/orchestrator/async-agents-group-wait.test.ts]: Fixed: removed timing assertion. Test now validates behavior (returns empty set) without wall-clock bounds.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `tests/node/orchestrator/async-agents-group-wait.test.ts:11` (qodo finding) — Test reimplements wait logic
> Addressed: Fixed: extracted waitForResultFiles into async-wait.ts. Test imports and exercises the real production helper.
### `extensions/orchestrator/async-agents.ts:268` (qodo bug) — Wait includes fire-and-forget
> Addressed: Fixed: missingJobs now filters out fireAndForget jobs.
### `tests/node/orchestrator/async-agents-group-wait.test.ts:75` (qodo bug) — Timing assertion can flake
> Addressed: Fixed: removed timing assertion. Test now validates behavior (returns empty set) without wall-clock bounds.
### `extensions/orchestrator/async-agents.ts:271` (qodo finding) — deliverGroupResults wait lacks test
> Addressed: Fixed: test imports real waitForResultFiles from async-wait.ts with 4 tests.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/async-wait.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+      if (found.has(id)) continue;
+      if (existsSync(join(resultDir, `${id}.json`))) {
+        found.add(id);
+      } else {
+        allFound = false;
+      }
+    }
+    if (allFound) break;
+    await new Promise(r => setTimeout(r, 250));
+  }
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/async-wait.ts]: Fixed: do-while loop scans at least once. Sleep uses Math.min(250, remaining). Added deadlineMs=0 test.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/async-agents.ts [hunk 14]
```diff
@@ -262,10 +263,15 @@ export function registerAsyncAgents(
...
-    // before processResultFile() has read all group members' outputs
+    // before processResultFile() has read all group members' outputs.
+    // Wait up to 2s total (not per-job) for all missing result files.
     const lateIngestedIds = new Set<string>();
-    for (const j of groupJobs) {
-      if (j.output !== undefined) continue; // Already ingested
+    const missingJobs = groupJobs.filter(j => j.output === undefined && !j.fireAndForget);
+    if (missingJobs.length > 0) {
+      await waitForResultFiles(ASYNC_RESULTS_DIR, missingJobs.map(j => j.id), 2000);
+    }
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/async-agents.ts]: Fixed: missingJobs excludes jobs with status failed so killed jobs do not delay group delivery.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/async-agents.ts:269` (qodo bug) — Killed job delays group delivery
> Addressed: Fixed: missingJobs excludes jobs with status failed so killed jobs do not delay group delivery.
### `extensions/orchestrator/async-wait.ts:18` (qodo bug) — Wait deadline can overshoot
> Addressed: Fixed: do-while loop scans at least once. Sleep uses Math.min(250, remaining). Added deadlineMs=0 test.
### `extensions/orchestrator/async-agents.ts:271` (qodo finding) — deliverGroupResults wait lacks test
> Addressed: Fixed: tests import real waitForResultFiles from async-wait.ts. 5 tests cover all behaviors.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/async-agents.ts [hunk 15]
```diff
@@ -262,10 +292,15 @@ export function registerAsyncAgents(
...
     const lateIngestedIds = new Set<string>();
-    for (const j of groupJobs) {
-      if (j.output !== undefined) continue; // Already ingested
+    const missingJobs = groupJobs.filter(j => j.output === undefined && !j.fireAndForget && j.status !== "failed");
+    if (missingJobs.length > 0) {
+      await waitForResultFiles(ASYNC_RESULTS_DIR, missingJobs.map(j => j.id), 2000);
+    }
+    for (const j of missingJobs) {
+      if (j.output !== undefined) continue;
       const rp = path.join(ASYNC_RESULTS_DIR, `${j.id}.json`);
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/async-agents.ts]: Fixed: late-ingest now includes failed jobs (reads their result files if they exist). Wait still excludes failed jobs (no point waiting). No result files are deleted without being ingested first.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/async-agents.ts:298` (qodo bug) — Failed results can be dropped
> Addressed: Fixed: late-ingest now includes failed jobs (reads their result files if they exist). Wait still excludes failed jobs (no point waiting). No result files are deleted without being ingested first.

---
*Assisted-by: PI (claude-opus-4-6-1m)*
